### ISSUE TITLE
Type hints are not respected for Enum members

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#1521]
**Describe the bug**
Type hints are not respected for Enum members

**To Reproduce**
```python
from enum import Enum

class MyEnum(Enum):
    ENTRY = (123, "abc")

    foo: int
    bar: str

    def __init__(self, foo: int, bar: str) -> None:
        self.foo = foo
        self.bar = bar

baz: int = 123 + MyEnum.ENTRY.foo
```

**Expected behavior**
No error

**Screenshots or Code**

```
$ pyright
Loading configuration file at /tmp/pyright3/root/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Linux
No include entries specified; assuming /tmp/pyright3/root
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
stubPath /tmp/pyright3/root/typings is not a valid directory.
Searching for source files
Found 1 source file
/tmp/pyright3/root/src/repro.py
  10:14 - error: Cannot assign member "foo" for type "MyEnum"
    Expression of type "int" cannot be assigned to member "foo" of class "MyEnum"
      "int" cannot be assigned to type "Literal[MyEnum.foo]" (reportGeneralTypeIssues)
  11:14 - error: Cannot assign member "bar" for type "MyEnum"
    Expression of type "str" cannot be assigned to member "bar" of class "MyEnum"
      "str" cannot be assigned to type "Literal[MyEnum.bar]" (reportGeneralTypeIssues)
  13:12 - error: Operator "+" not supported for types "Literal[123]" and "Literal[MyEnum.foo]"
    Operator "+" not supported for types "Literal[123]" and "Literal[MyEnum.foo]" (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 infos
Completed in 0.612sec
```

**VS Code extension or command-line**
```
$ pyright --version                                                                                         1
pyright 1.1.109
```

**Additional context**
* This seems to work fine with mypy.
* I tried declaring the members before the enum entries, but it didn't change anything.

**Comment by erictraut** [comment]
Thanks for the bug report.

Mypy doesn't generate any errors, but it is also not handling this correctly. It appears to think that `foo` and `bar` are entries within the enumeration, not instance variables within the enum instances.

Try adding the following lines to your sample and run mypy.

```python
reveal_type(MyEnum.ENTRY)
reveal_type(MyEnum.foo)
```

There appears to be an ambiguity in the typing specs about how to interpret `foo: int` in this case.

I've posted a [question to the typing-sig mailing list](https://mail.python.org/archives/list/typing-sig@python.org/thread/NTCDIWE3QUM4JLISJN4ZVXAH6JX2T6TL/) to see if the maintainers of typeshed, mypy, or any of the other type checkers can offer insights.

**Comment by ipkiss42** [comment]
Thanks the super quick investigation!

I tested the workaround proposed in response to your post and it works fine with Pyright.

So there are two workarounds, if changing the code is an option:
1. Do not specify enum fields explicitly, relying indirectly on `__init__()`.
2. Make the enum inherit from a `dataclass` declaring the fields, to avoid having to write the `__init__()` manually.

**Comment by erictraut** [comment]
It appears that there is no convention defined in PEP 484 or elsewhere for correctly annotating an enum class so members of the enumeration can be distinguished from instance variables within each enum instance. For now, I've changed the behavior within pyright as follows:

Within a stub (".pyi") file, it uses these rules:

* Any variable declared within an enum class is considered a member of the enumeration unless it is named `value`, `name` or a dundered name (which are reserved by the EnumMeta metaclass).

Within a non-stub (".py") file, it uses these rules:

* A variable declared within an enum class declaration is considered a member of the enumeration only if it is part of an assignment statement.
* A variable declared with an enum class with a "bare" type annotation and no assignment is treated as an instance variable.

If the typing community comes to some other consensus, I'll update pyright accordingly. I don't think that's likely to happen any time soon though. There are too many stubs and type checkers that already assume the rules above for stubs.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.114, which I just published. It will also be in the next release of Pylance.

**Comment by ipkiss42** [comment]
Thank you! Your approach seems like a good tradeoff to me.
